osu.Game/Screens/Play/SubmittingPlayer.cs

```diff
@@ -0,0 +1,120 @@
+            var tcs = new TaskCompletionSource<bool>();
+
+            if (!api.IsLoggedIn)
+            {
+                fail(new InvalidOperationException("API is not online."));
+                return;
+            }
+
+            var req = CreateTokenRequestRequest();
+
+            if (req == null)
+            {
+                fail(new InvalidOperationException("Request could not be constructed."));
+                return;
+            }
+
+            req.Success += r =>
+            {
+                Token = r.ID;
+                tcs.SetResult(true);
+            };
+            req.Failure += fail;
+
+            api.Queue(req);
+
+            tcs.Task.Wait();
+
+            void fail(Exception exception)
+            {
+                if (HandleTokenRetrievalFailure(exception))
+                {
+                    if (string.IsNullOrEmpty(exception.Message))
+                        Logger.Error(exception, "Failed to retrieve a score submission token.");
+                    else
+                        Logger.Log($"You are not able to submit a score: {exception.Message}", level: LogLevel.Important);
+
+                    Schedule(() =>
+                    {
+                        ValidForResume = false;
+                        this.Exit();
+                    });
+                }
+
+                tcs.SetResult(false);
+            }
+
+            base.LoadAsyncComplete();
+        }
+
+        /// <summary>
+        /// Called when a token could not be retrieved for submission.
+        /// </summary>
+        /// <param name="exception">The error causing the failure.</param>
+        /// <returns>Whether gameplay should be immediately exited as a result. Returning false allows the gameplay session to continue. Defaults to true.</returns>
+        protected virtual bool HandleTokenRetrievalFailure(Exception exception) => true;
+
+        protected override async Task PrepareScoreForResultsAsync(Score score)
+        {
+            await base.PrepareScoreForResultsAsync(score).ConfigureAwait(false);
+
+            // token may be null if the request failed but gameplay was still allowed (see HandleTokenRetrievalFailure).
+            if (Token == null)
+                return;
+
+            var tcs = new TaskCompletionSource<bool>();
+            var request = CreateSubmissionRequest(score, Token.Value);
+
+            request.Success += s =>
+            {
+                score.ScoreInfo.OnlineScoreID = s.ID;
+                tcs.SetResult(true);
+            };
+
+            request.Failure += e =>
+            {
+                Logger.Error(e, "Failed to submit score");
+                tcs.SetResult(false);
+            };
+
+            api.Queue(request);
+            await tcs.Task.ConfigureAwait(false);
+        }
+
+        protected abstract APIRequest<MultiplayerScore> CreateSubmissionRequest(Score score, long token);
```

Comment:
Got slightly confused why this referenced `MultiplayerScore` at this level already instead of `RoomSubmittingPlayer` when reading. That class doesn't seem to be overly related to multiplayer specifically. May be better to rename it to `OnlineScore`?

Reply:
it'll probably change going forward, yep.